### PR TITLE
 Add function pushoradd!

### PR DESCRIPTION
**Pull request by baggepinnen**
`pushoradd!` is useful if, e.g., one is adding to a timer incrementally
I found the need for this function, maybe others will as well

### CODE REVIEW

File: src/mvhistory.jl [hunk 1]
```diff
@@ -1,3 +1,5 @@
+export pushoradd!
```
Comment (Evizero) [inline src/mvhistory.jl]: Let's keep the `export` statements centralized in the main file.

File: src/mvhistory.jl [hunk 2]
```diff
@@ -97,3 +99,16 @@ macro trace(tr, i, vars...)
...
 end
+
+"""
+    pushoradd!(trace, key, iter, val)
+
+Increments the value for a given key and iteration if it exists, otherwise adds the key/iteration pair with an ordinary push.
+"""
+function pushoradd!(trace::MultivalueHistory, key::Symbol, iter::Number, val)
+    if haskey(trace, key) && length(trace.storage[key]) >= iter
+        trace[key].values[iter] += val
```
Comment (Evizero) [inline src/mvhistory.jl]: Accessing the member variable `values` is rather specific to `MultivalueHistory{<:History}` and would not work with `QHistory`. I am ok with that but we should catch this earlier I think. maybe dispatch the whole method on `trace::MultivalueHistory{<:History}`.

File: src/mvhistory.jl [hunk 3]
```diff
@@ -97,3 +99,16 @@ macro trace(tr, i, vars...)
...
     block
 end
+
+"""
+    pushoradd!(trace, key, iter, val)
+
+Increments the value for a given key and iteration if it exists, otherwise adds the key/iteration pair with an ordinary push.
+"""
+function pushoradd!(trace::MultivalueHistory, key::Symbol, iter::Number, val)
+    if haskey(trace, key) && length(trace.storage[key]) >= iter
```
Comment (Evizero) [inline src/mvhistory.jl]: `length(trace.storage[key]) >= iter`

i don't think that works in the general case. One can store disconnected iterations that don't follow an explicit system other than being ordered (e.g. the iterations `1, 10, 12, 100`).  So i think to find the actual index of the associated value of `iter` one would have to use something like 

```julia
i = findfirst(trace.storage[key].iterations, iter)
```

with that `i` one can then check if `i !=0 ` (i.e. the element exists) and if it does use 

```julia
trace[key].values[i] += val
```

**Review by Evizero — CHANGES_REQUESTED** [review]
This seems like a useful addition. Aside from the three individual remarks, could you maybe add some basic tests for the (i think) 4 different scenarios that could occur here 

- `MethodError` when called with a multivalue history of a `QHistory`
- key does not yet exist
- key exists but iteration was not yet stored
- key and iteraton exist already

optionally a mention in the readme would be nice

File: src/mvhistory.jl (hunk 2 above)
Comment (baggepinnen) [reply src/mvhistory.jl]: I can't seem to figure out how to create an `MVHistory{QHistory}`?
Also, `trace::MultivalueHistory{<:History}` Gives me a type error, I can only get it to work with `trace::MVHistory{<:History}`

Comment (Evizero) [reply src/mvhistory.jl]: > I can't seem to figure out how to create an MVHistory{QHistory}?

This constructor should work?: https://github.com/JuliaML/ValueHistories.jl/blob/master/src/mvhistory.jl#L5

> Also, trace::MultivalueHistory{<:History} Gives me a type error, I can only get it to work with trace::MVHistory{<:History}

right, my bad. The later then

**Comment by baggepinnen** [comment]
This should be complete now, apart from the fact that I used `trace::MVHistory{<:History}` instead of `trace::MultivalueHistory{<:History}` as commented above.

**Review by Evizero — APPROVED** [review]

**Comment by Evizero** [comment]
Looks good to me. thanks for doing this.

I'll merge after CI passes. Unfortunately I am travelling all day, so this may not be before tomorrow

**Comment by baggepinnen** [comment]
Great! Thanks for the feedback. I have been using this package for a lot of things, nice to be able to contribute =)

**Pull request merged**